### DISCUSSION TITLE
Improve load strategy JOINED

### DISCUSSION THREAD

**Discussion by angelvega93** [mikro-orm/mikro-orm#4610] [Ideas]
Hi, a few days ago DrizzleORM updated the way they fetch relational queries in a single query, it would be good if mikroorm could fetch relational queries this way too, the current method is slow because it can't use the indexes correctly and is needed to use the strategy "SELECT_IN" to use indexes correctly.

https://github.com/drizzle-team/drizzle-orm/releases/tag/0.28.0

**Comment by B4nan** [comment]
If you have a perf issue, please provide a complete repro and we can talk about how to make it fast, I don't plan to change anything in the joined strategy currently, not without an actual problem I would be solving.

**Reply by angelvega93** [reply]
This is an example of the generated query with joined strategy in photos relation in my current project:
```sql
select 
  "p0"."id", 
  "p0"."title", 
  "p0"."description", 
  "p0"."status", 
  "p0"."created_at", 
  "p0"."updated_at", 
  "p0"."weight",
  "p0"."approved",
  "p1"."id" as "p1__id", 
  "p1"."order" as "p1__order", 
  "p1"."uri" as "p1__uri", 
  "p1"."post_id" as "p1__post_id", 
  "p1"."created_at" as "p1__created_at", 
  "c2"."id" as "c2__id", 
  "c2"."label" as "c2__label" 
from 
  "post" as "p0" 
  left join "photo" as "p1" on "p0"."id" = "p1"."post_id" 
  left join "category" as "c2" on "p0"."category_id" = "c2"."id" 
where 
  "p0"."id" in (
    select 
      "p0"."id" 
    from 
      (
        select 
          "p0"."id" 
        from 
          "post" as "p0" 
          left join "photo" as "p1" on "p0"."id" = "p1"."post_id" 
          left join "category" as "c2" on "p0"."category_id" = "c2"."id" 
        where 
          "p0"."status" = 1 
          and "p0"."approved" = true 
        group by 
          "p0"."id" 
        order by 
          min("p0"."weight") desc, 
          min("p1"."order") asc 
        limit 
          20
      ) as "p0"
  ) 
order by 
  "p0"."weight" desc, 
  "p1"."order" asc
```

And this is the query plan:
![dbeaver_Vm6vBd7sHt](https://github.com/mikro-orm/mikro-orm/assets/9835055/2b6ea36f-9fd8-4225-8b93-34938ee13313)


Now if i change strategy to SELECT_IN on photos relation, the generated queries are the following:
```sql
select 
  "p0"."id", 
  "p0"."title", 
  "p0"."description", 
  "p0"."status", 
  "p0"."approved",
  "p0"."weight",
  "p0"."created_at", 
  "p0"."updated_at", 
  "c1"."id" as "c1__id", 
  "c1"."label" as "c1__label" 
from 
  "post" as "p0" 
  left join "category" as "c1" on "p0"."category_id" = "c1"."id" 
where 
  "p0"."status" = 1 
  and "p0"."approved" = true 
order by 
  "p0"."weight" desc 
limit 
  20
```
![dbeaver_LJSAQrJq2I](https://github.com/mikro-orm/mikro-orm/assets/9835055/5f136ef0-82cc-4d55-8820-c7fcef67afde)

```sql
select 
  "p0".* 
from 
  "photo" as "p0" 
where 
  "p0"."post_id" in (
    316, 425, 182, 301, 346, 409, 112, 127, 
    207, 72, 318, 322, 107, 362, 29, 111, 
    236, 116, 203, 496
  ) 
order by 
  "p0"."order" asc, 
  "p0"."post_id" asc
```

![image](https://github.com/mikro-orm/mikro-orm/assets/9835055/1d65005b-2a13-4b1f-b802-b1834c3460ac)

If i fill the database with dummy data it gets slower and slower compared to SELECT_IN.

**Reply by B4nan** [reply]
So we are talking about pagination, not about the joined strategy in general. If you can provide a complete repro, I can take a look, but this is usually about missing indexes (there was a recent report about this as well, with the same resolution - missing index). 

Did you try the same with drizzle and you see better times in there? How does the query there look like? 

---

All I am trying to do here is to push you to propose an actual feature request, I don't use drizzle, and a link to their changelog is not helping me much in terms of understanding what you are actually after... With that said, the pagination mechanism is a pretty complex thing, and doing some changes in there can be painful and can have wild side effects.

**Reply by angelvega93** [reply]
This is the generated query by drizzle:
```sql
select 
  "post"."id", 
  "post"."title", 
  "post"."description", 
  "post"."status", 
  "post"."created_at", 
  "post"."updated_at", 
  "post"."weight", 
  "post"."approved", 
  "post"."category_id", 
  "post_photos"."data" as "photos", 
  "post_category"."data" as "category" 
from 
  "post" 
  left join lateral (
    select 
      coalesce(
        json_agg(
          json_build_array(
            "post_photos"."id", "post_photos"."order", 
            "post_photos"."uri", "post_photos"."title", 
            "post_photos"."published", "post_photos"."post_id", 
            "post_photos"."created_at"
          )
        ), 
        '[]' :: json
      ) as "data" 
    from 
      "photo" "post_photos" 
    where 
      "post_photos"."post_id" = "post"."id"
  ) "post_photos" on true 
  left join lateral (
    select 
      json_build_array(
        "post_category"."id", "post_category"."label"
      ) as "data" 
    from 
      (
        select 
          * 
        from 
          "category" "post_category" 
        where 
          "post_category"."id" = "post"."category_id" 
        limit 
          1
      ) "post_category"
  ) "post_category" on true 
limit 
  20
```

And this is the query planner:
![image](https://github.com/mikro-orm/mikro-orm/assets/9835055/76943733-e6fb-4b3b-af9d-c8a175eafdf9)

**Reply by B4nan** [reply]
Interesting, I'd be curious to see how this works in other drivers, not sure if you can do the same kind of magic with e.g. mysql.

Could you provide a repro for this so I can experiment myself? Ideally including the drizzle setup as well so I can compare things side by side.

Sidenote: I've been also playing around with the idea of changing the backends - replacing knex with something else, and drizzle is one of the candidates. This will be a major task and it won't fit into v6 for sure, but for v7 I'd like to get rid of knex one way or the other (the current plan is more about having a custom implementation as that feels more straightforward and futureproof, but having something like drizzle would unlock the possibility of new drivers, as well as supporting deno).

**Reply by angelvega93** [reply]
Ok, i will try to have a simple demo repository tomorrow.

**Reply by B4nan** [reply]
Btw that drizzle changelog mentions this as the very first thing:

> Removed support for filtering by nested relations

Which they further comment on as something they had to do to better optimize the queries. This is a big no-no for me, filtering by relations will definitely need to stay working as it is now, I am quite sure that removing support for that would piss off many users :] But maybe I am reading this too much in between the lines, as said, I have no experience with that library and will be happy to take some inspiration from their approach if it's feasible for us.

**Reply by angelvega93** [reply]
After some tests, the problem is the loading strategy JOINED if using ORDER BY.

Here is an example repo:
https://github.com/angelvega93/mikro-orm-drizzle
```
npm install
npx mikro-orm schema:fresh --run --seed
npm run start
```

Another advantage of using laterals is its efficient use of bandwidth, as it avoids data repetition, it would be great if we can have another loading strategy "LATERAL", this will prevent modifying the existing "JOINED" strategy.

**Reply by B4nan** [reply]
Well, too bad knex does not support lateral joins (https://github.com/knex/knex/issues/3732), so we would have to implement that completely ourselves.

**Reply by B4nan** [reply]
> the problem is the loading strategy JOINED if using ORDER B

Can you try v6? I've been doing some internal refactorings and ended up removing the default ordering for joined relations, so maybe it's about that?

**Reply by angelvega93** [reply]
I updated to v6, the generated query have the same problem with "orderBy".

**Reply by B4nan** [reply]
Hmm looking at the query from drizzle, there is no `order by` clause, isn't that a bug on their end? How could it order things (which you [specify explicitly](https://github.com/angelvega93/mikro-orm-drizzle/blob/main/src/main.ts#L26)) without that?

I thought the problem was with the implicit ordering, which is removed in v6, but you [specify the ordering explicitly](https://github.com/angelvega93/mikro-orm-drizzle/blob/main/src/mikro-orm/entities/post.entity.ts#L42).

**Reply by angelvega93** [reply]
The `order by` is working correctly in drizzle, i don't know why it allowed me to put `schema.post.id` instead of `schema.comment.id`, also i miss the `desc` function to match the mikroorm order.

![image](https://github.com/mikro-orm/mikro-orm/assets/9835055/51c1a895-ba12-4e50-90ec-65897a78c87b)

I guess they could omit the order by marked in blue.

**Reply by B4nan** [reply]
Ok, so you were comparing different queries. Is the new query from drizzle also performant?

**Reply by angelvega93** [reply]
The problem is only when using order by on MikroORM, the generated query is slow.
